source/common/network/listen_socket_impl.h

```diff
@@ -80,6 +80,28 @@ template <typename T> class NetworkListenSocket : public ListenSocketImpl {
 
   Socket::Type socketType() const override { return T::type; }
 
+  // These four overrides are introduced to perform check. A null io handle is possible only if the
+  // io handle is listener side.
+  IoHandle& ioHandle() override {
+    ASSERT(io_handle_ != nullptr);
+    return *io_handle_;
+  }
+  const IoHandle& ioHandle() const override {
+    ASSERT(io_handle_ != nullptr);
+    return *io_handle_;
+  }
+  void close() override {
+    if (io_handle_ != nullptr) {
```

Comment:
In working through my other change, I'm hitting a similar issue on `setSocketOption` (now that reuse port is the default). I think there is a similar bug there also. It makes me think we are going about the no bind sockets incorrectly in general if we need all these guards.

Reply:
(i.e. why do no bind listeners even have sockets. I realize this is a much larger change and I'm trying to get through my current mess first, but this seems like something we might want to fix in the future.)

Reply:
SetSocketOption should check bind_to_port w/ or w/o PR.  At least ListenSocketImpl::setupSocket has a bind_to_port arg.

> why do no bind listeners even have sockets

Good point! I was thinking about create a derived not-bind-listen-socket and you think deeper